src/main/java/duke/DukeController.java

```diff
@@ -0,0 +1,245 @@
+        printNewLine();
+    }
+    public void saveData() throws IOException {
+        File dataDirectory = new File(DATA_DIRECTORY);
+        if (!dataDirectory.exists()) {
+                boolean isDirectoryCreated = dataDirectory.mkdir();
+                if (!isDirectoryCreated) {
+                    System.out.println(CREATE_DIRECTORY_FAILED);
+                    printNewLine();
+                    return;
+                }
+        }
+        File saveFile = new File(savedFilePath.toString());
+        FileWriter fileWriter = new FileWriter(saveFile);
+        for (Task task:taskList) {
+            String taskDetails = task.getSaveString();
+            fileWriter.write(taskDetails);
+            fileWriter.write('\n');
+        }
+        fileWriter.close();
+        System.out.println(SAVE_FILE_SUCCESS);
+        printNewLine();
+    }
+
+    private final InputParser inputParser = new InputParser();
+
+    public void getUserInput(){
+        userInput = inputParser.readInput();
+    }
+    public void handleUserInput(){
+        try {
+            String[] parsedInput = inputParser.parseInput(userInput);
+            String userCommand = parsedInput[0];
+            switch(userCommand){
+            case "bye":
+                printGoodbye();
+                Duke.exitDuke();
+                return;
+            case "list":
+                printTaskList();
+                break;
+            case "todo":
+            case "deadline":
+            case "event":
+                try {
+                    String[] details = inputParser.parseTaskInformation(parsedInput[1], userCommand);
+                    Task newTask = generateTask(userCommand, details);
+                    addTask(newTask);
+                    saveData();
+                } catch (IllegalInputException e) {
+                    System.out.println(e.getErrorMessage());
+                    printNewLine();
+                } catch (IOException e){
+                    System.out.println(e.getMessage());
+                    printNewLine();
+                }
+                break;
+            case "delete":
+            case "mark":
+            case "unmark":
+                try {
+                    int taskIndex = inputParser.parseTaskIndex(parsedInput[1], userCommand);
+                    if (userCommand.equals("mark") || userCommand.equals("unmark")) {
+                        updateTaskStatus(taskIndex, userCommand);
+                    } else {
+                        deleteTask(taskIndex);
+                    }
+                    saveData();
+                } catch (IllegalInputException e){
+                    System.out.println(e.getErrorMessage());
+                    printNewLine();
+                } catch (IOException e){
+                    System.out.println(e.getMessage());
+                    printNewLine();
+                }
+                break;
+            default:
+                printNewLine();
```

Comment:
[Code Quality] The logic inside each case could be extracted out for better SLAP and control flow.